osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs

```diff
@@ -83,7 +83,7 @@ private DifficultyAttributes calculate(IBeatmap beatmap, Mod[] mods, double cloc
                     foreach (Skill s in skills)
                     {
                         s.SaveCurrentPeak();
-                        s.StartNewSectionFrom(currentSectionEnd);
+                        s.StartNewSectionFrom(currentSectionEnd / clockRate);
```

Comment:
This seems wrong -
1. The outer loop and the initialisation of `currentSectionEnd` use `HitObject.StartTime` and not the clock adjusted start time.
2. `currentSectionEnd` is already clock-adjusted albeit seemingly in the wrong direction from what you're expecting - see above where the clock is factored into `sectionLength`. So for a base length of `400` we get `400 * 1.5 / 1.5 = 400 = DT`
3. The placement of this is off - it should be factored into `sectionLength` rather than post-divided.

Reply:
> This seems wrong

I totally agree, however unless I messed something up in my testing, it isn't.
The root cause of this bug is that `currentSectionEnd` is not a clockrate adjusted value (since it is ultimately based on `HitObject.StartTime`) and `Skill.StartNewSectionFrom()` expects a clockrate adjusted value, therefore we must adjust it manually.

This could *probably* be rewritten to be more clear, however I decided not to do this for two reasons:
1. It's a 1 line change to fix the actual bug without rewriting
2. This all gets refactored anyway in the dependent PR https://github.com/ppy/osu/pull/11689 which simplifies all of this

This change can be somewhat verified by the fact that it's values are consistent with the those in the refactor in https://github.com/ppy/osu/pull/11689 which rewrites all of this in a way that this bug wouldn't be possible.

Reply:
As far as I can understand, this code indeed seems to be correct, because:

* the virtual public `SectionLength` is real/wall clock time
* the local `sectionLength` is track time (converted by multiplying by rate - this variable has a bad/confusing name)
* `currentSectionEnd` is track time (uncompensated hitobject start time divided by the local section length)
* `StartNewSectionFrom()` is real/wall clock time (divided back by rate again; `GetPeakStrain()` lower down depends on it, as it uses difficulty hit object start times, which are compensated for time)

That said, this *is* mightily confusing. So unless the refactor changes things so that it is absolutely clear where track time goes and where real time goes, I'd add some comments so that future readers don't get lost.

Reply:
Yep that's exactly right.
The refactor removes the concept of base timings completely from this process so it's much more simplified.
Any transformations are done in preprocessing and then everything is handled uniformly from there.